x-pack/plugin/shutdown/src/main/java/org/elasticsearch/xpack/shutdown/ShutdownService.java

```diff
@@ -0,0 +1,85 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.shutdown;
+
+import org.apache.log4j.LogManager;
+import org.apache.log4j.Logger;
+import org.apache.logging.log4j.message.ParameterizedMessage;
+import org.elasticsearch.cluster.ClusterChangedEvent;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.ClusterStateListener;
+import org.elasticsearch.cluster.ClusterStateUpdateTask;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.cluster.metadata.NodesShutdownMetadata;
+import org.elasticsearch.cluster.metadata.SingleNodeShutdownMetadata;
+import org.elasticsearch.cluster.service.ClusterService;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.function.Function;
+import java.util.stream.Collectors;
+
+public class ShutdownService implements ClusterStateListener {
+    private static final Logger logger = LogManager.getLogger(ShutdownService.class);
+
+    final ClusterService clusterService;
+
+    public ShutdownService(ClusterService clusterService) {
+        this.clusterService = clusterService;
+        clusterService.addListener(this);
+    }
+
+    @Override
+    public void clusterChanged(ClusterChangedEvent event) {
+        if (event.state().nodes().isLocalNodeElectedMaster() == false) {
+            // Only do this if we're the current master node.
+            return;
+        }
```

Comment:
If you wanted, you could also add another check for

```java
if (event.nodesAdded() == false) {
    return;
}
```

To only perform the check when a new node has been added to the cluster?

Reply:
I'm kind of worried what would happen to an overloaded cluster if we don't have this check, for example:

- node is added, CS task is submitted
- a different state is received (some other change), CS task is submitted again (because it hasn't been updated yet
- a different state is received (some other change), task is submitted again
- a different state is received (some other change), task is submitted again
- the first CS task runs and updates the state
- the second CS task runs and updates the state (even though there are no new nodes)
- the third CS task runs and updates the state (even though there are no new nodes)

The `if` check I suggested below would help, but we may be able to avoid submitting the task if we only submit when a node has been added.